### ISSUE TITLE
Handling of files in the `content` field of `PbPage` model

### ISSUE THREAD

**Issue by doitadrian** [webiny/webiny-js#614]
## This is:
- Idea

## Actual Behavior

Before we start off, let's give some background.

## Background
In PageBuilder, we have `PpPage` model and in it, the `content` field.

The `content` field may store images. The following is an excerpt in
which we can see how a file is saved in the database.

```
(...)
    {
      "id": "Of2JQrobK",
      "data": {
        "image": {
          "file": "5db3410646e6da8d61cc1393"
        }
      },
      "elements": [],
      "path": "0.0.0.0.2",
      "type": "image"
    }
(...)
```

This value of `file` key here is a result of several steps:
1. In the Page Builder, user picks an image from File Manager. Note that the File Manager doesn't return the file's ID, but a full data object: `{ src, size, id, type, key }`. So once the client sends the revision update mutation, this whole object is sent.

2. On the backend, we have a set of plugins that completely traverse received `content`, and do the replacing upon saving to the database. So instead of the full received object, only `id` is saved, like in the first example shown above.

3. Again, when fetching data from the database, we have the opposite process, in which a set of plugins convert these IDs into `{ src, id }` objects, so that the frontend again has the initial structure od data. Note that only `src` and `id` values are returned, no need really for other keys. But it can be added if needed.

## The problem
The problem is currently happening in step 3. Currently, when we want to replace the ID with the object, we need to load the file, and we do that by directly accessing the database connection, and doing a query on `File` collection, which is not an ideal approach.

The second approach that was tried here was doing a GQL query on `api-files` service, which is much better in terms of microservices approach. But this time, the problem is that getting/listing pages becomes slow since each page has to do an HTTP request (1 page = 1 HTTP request that fetched multiple files). We have an even worse case when you do a `listPages`, in which you might query for 10 or more pages. A lot of HTTP requests and a lot of latency is being added.

So is there a third way that we could solve this?

## Possible Solution
To solve the above-explained problem, we could save files a bit differently.

Instead of saving the file `id` to the database, we would save file `key`. This is good because field `key` never changes and since we have it, we can easily construct the full URL to the file upon receiving the page data from the database.

We would only need to fetch the prefix for all files, which is stored in `api-files` service's settings, and which can be fetched via GQL. This would need to be fetched only once so it's not super intensive on the backend.

So to summarize, this would free us from making GQL requests (fetching all files by their IDs), since we would be able to construct full paths directly in code.

### Problem - File Manager File-Replacement Feature 
Question: do we want a file-replacement feature in the File Manager? In other words, can a user have a single image selected in multiple pages, and then go to the File Manager and choose that same image and replace it with another one? So basically `id` of the file stays the same, and the `key` to which it points changes.

If so, the above solution won't work, since it proposes saving file `key` instead of `id`. And changing the key of a single file via File Manager would not apply changes to the `content` field in pages.

**Comment by Pavel910** [comment]
Whatever we decide in the end, I'd suggest we store an `object`, and not a plain `string` value. Something like:

```js
{ __type: "file", id: "1234567", "key": "/234234-file-name.png" }
```

This gives us space for future upgrades if the need arises. It also allows you to easily detect files in the `content`. You simply traverse the content object recursively looking for `__type` key so no modifier plugins are needed.

**Comment by doitadrian** [comment]
@Pavel910 and I discussed the "Problem - File Manager File-Replacement Feature " just now.

We both agreed that this feature is not necessary and maybe kinda unreliable since the user might not be aware of all changes that his action might create.

We both agreed that, in the context of the Page Builder, if a user wants to use the same image on several pages, that he'll create a global element that will contain that image. Now, if there's a need to change it, he'll select a different image in that global element, not change the file via File Manager.

This all means that the proposed solution, along with the additional comment from @Pavel910 should work well.

🚀 🚀 🚀

**Comment by Pavel910** [comment]
Just to clarify, a `global` element is only required if you know you will want to change the image across multiple pages in the future. If you don't need that, you can insert the same image from File Manager across different pages. Since the File Manager itself will not allow file content replacement  (you can't upload a different file in place of the existing file), you are safe to reuse existing images across the system.

**Comment by doitadrian** [comment]
Please reopen this issue once you add more information and updates here. If this is not the case and you need some help, feel free to seek help from our [Slack Community](https://www.webiny.com/slack) or ping one of the reviewers. Thank you for your contributions!
